hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/SystemContractMethod.java

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.exec.utils;
+
+import static java.util.Objects.requireNonNull;
+
+import com.esaulpaugh.headlong.abi.Function;
+import com.esaulpaugh.headlong.abi.Tuple;
+import com.esaulpaugh.headlong.abi.TupleType;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.EnumSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes;
+
+/**
+ * Represents a system contract method: It's signature + outputs, and other information about how it is declared
+ * @param function A headlong description of this method: signature + outputs, selector
+ * @param systemContract Which system contract this method is part of (HAS, HSS, HTS, PRNG, EXCHANGE)
+ * @param via Whether called directly or via redirect (aka proxy)
+ * @param categories Set of category flags possibly empty
+ * @param modifier An optional Solidity function modifier (e.g., pure or view)
+ * @param variants Set of method variants (e.g., version number, or FT vs NFT), possibly empty
+ */
+public record SystemContractMethod(
+        @NonNull Function function,
+        @NonNull Optional<SystemContract> systemContract,
+        @NonNull CallVia via,
+        @NonNull EnumSet<Category> categories,
+        @NonNull Optional<Modifier> modifier,
+        @NonNull EnumSet<Variant> variants) {
```

Comment:
Seems reasonable.  May want to add a system contract dimension as well for system contract versioning work later

Reply:
Agree, but not sure what it would look like.  Are we going to _name_ versions, e.g. with an enum, or just refer to them by address, or ???  I don't think we have an epic for system contract versioning yet, do we?  I'd put this improvement as a task in that epic (it'll be necessary at that time as you say).